File: src/engine/youtube/formatSelector.ts
/**
 * YouTube Format Selector
 *
 * Selects the best video and audio formats from available options.
 */

import { Logger } from "../../core/logger.js";
import { ConfigManager } from "../../core/config.js";
import type { Format } from "../../types/youtube.js";

/**
 * Selected formats for download
 */
export interface SelectedFormats {
  video: Format | null;
  audio: Format | null;
}

/**
 * Options for manual format selection
 */
export interface FormatSelectionOptions {
  selectedVideoItag?: number;   // User-chosen video format itag
  selectedAudioItag?: number;   // User-chosen audio format itag
}

/**
 * Format Selector
 */
export class FormatSelector {
  private logger: Logger;

  constructor() {
    this.logger = Logger.getInstance();
  }

  /**
   * Select formats with optional manual itag selection.
   * If a selected itag is found in the available formats, it is used.
   * Otherwise falls back to automatic best-quality selection.
   */
  selectWithOptions(formats: Format[], options?: FormatSelectionOptions): SelectedFormats {
    let video: Format | null = null;
    let audio: Format | null = null;

    // Try manual video selection
    if (options?.selectedVideoItag != null) {
      video = formats.find(
        (f) => f.itag === options.selectedVideoItag && f.mimeType?.includes("video") && f.url,
      ) || null;
      if (video) {
        this.logger.debug(
          `[FormatSelector] Manual video selection: itag ${video.itag} ${video.width}x${video.height}@${video.fps || "?"}fps`,
        );
      } else {
        this.logger.warn(
          `[FormatSelector] Manual video itag ${options.selectedVideoItag} not found, falling back to auto`,
        );
      }
    }

    // Try manual audio selection
    if (options?.selectedAudioItag != null) {
      audio = formats.find(
        (f) => f.itag === options.selectedAudioItag && f.mimeType?.includes("audio") && f.url,
      ) || null;
      if (audio) {
        this.logger.debug(
          `[FormatSelector] Manual audio selection: itag ${audio.itag} ${audio.bitrate}bps`,
        );
      } else {
        this.logger.warn(
          `[FormatSelector] Manual audio itag ${options.selectedAudioItag} not found, falling back to auto`,
        );
      }
    }

    // Fall back to auto for any unresolved selections
    // A selectedItag of -1 means explicitly "none" (video-only or audio-only)
    if (options?.selectedVideoItag === -1) {
      video = null; // Explicitly no video
    }
    if (options?.selectedAudioItag === -1) {
      audio = null; // Explicitly no audio
    }

    const auto = this.selectBest(formats);
    if (!video && options?.selectedVideoItag !== -1) video = auto.video;
    if (!audio && options?.selectedAudioItag !== -1) audio = auto.audio;

    return { video, audio };
  }

  /**
   * Select the best video and audio formats
   */
  selectBest(formats: Format[]): SelectedFormats {
    const config = ConfigManager.getInstance().get();
    const maxRes = config.downloader.max_video_resolution || 9999;
    const prefer60 = config.downloader.prefer_60fps ?? true;

    let bestVideo: Format | null = null;
    let bestAudio: Format | null = null;

    // Debug: log all available video formats with URLs
    const videoFormatsWithUrls = formats.filter(f => f.mimeType?.includes("video") && f.url);
    if (videoFormatsWithUrls.length > 0) {
      this.logger.debug(
        `[FormatSelector] Available video formats with URLs: ${videoFormatsWithUrls.map(f => `itag ${f.itag} (${f.width}x${f.height}@${f.fps || "?"}fps, ${f.bitrate}bps)`).join(", ")}`
      );
    }

    for (const f of formats) {
      const mimeType = f.mimeType || "";

      if (mimeType.includes("video")) {
        // Only consider formats with URLs
        if (!f.url) continue;

        // Check resolution limit using max(width, height) to handle portrait videos
        const maxDimension = Math.max(f.width || 0, f.height || 0);
        if (maxDimension > maxRes) continue;

        if (!bestVideo) {
          bestVideo = f;
          continue;
        }

        // Prefer higher resolution within limit
        const currentMaxDim = Math.max(
          bestVideo.width || 0,
          bestVideo.height || 0,
        );

        if (maxDimension > currentMaxDim) {
          bestVideo = f;
          continue;
        }

        if (maxDimension === currentMaxDim) {
          // Same resolution — fps preference is tiebreaker
          const fFps = f.fps || 30;
          const bestFps = bestVideo.fps || 30;

          if (fFps !== bestFps) {
            if (prefer60 && fFps > bestFps) bestVideo = f;
            if (!prefer60 && fFps < bestFps) bestVideo = f;
            continue;
          }

          // Same fps — fall through to bitrate
          if (f.bitrate > bestVideo.bitrate) {
            bestVideo = f;
          } else if (f.bitrate === bestVideo.bitrate) {
            // Same bitrate — prefer lower auth level
            if ((f.authLevel ?? 999) < (bestVideo.authLevel ?? 999)) {
              bestVideo = f;
            }
          }
        }
      } else if (mimeType.includes("audio")) {
        // Only consider formats with URLs
        if (!f.url) continue;

        // Prefer higher bitrate audio, with authLevel as tiebreaker
        if (!bestAudio || f.bitrate > bestAudio.bitrate) {
          bestAudio = f;
        } else if (f.bitrate === bestAudio.bitrate && (f.authLevel ?? 999) < (bestAudio.authLevel ?? 999)) {
          bestAudio = f;
        }
      }
    }

    if (bestVideo) {
      this.logger.debug(
        `[FormatSelector] Best video: ${bestVideo.width}x${bestVideo.height}@${bestVideo.fps || "?"}fps @ ${bestVideo.bitrate} (itag ${bestVideo.itag}, source: ${bestVideo.source || "unknown"})`,
      );
    }
    if (bestAudio) {
      this.logger.debug(
        `[FormatSelector] Best audio: ${bestAudio.bitrate} (itag ${bestAudio.itag}, source: ${bestAudio.source || "unknown"})`,
      );
    }

    return { video: bestVideo, audio: bestAudio };
  }

  /**
   * Filter formats to only those with direct URLs
   */
  filterWithUrls(formats: Format[]): Format[] {
    return formats.filter((f) => !!f.url);
  }

  /**
   * Get video formats only
   */
  getVideoFormats(formats: Format[]): Format[] {
    return formats.filter((f) => f.mimeType?.includes("video") && f.url);
  }

  /**
   * Get audio formats only
   */
  getAudioFormats(formats: Format[]): Format[] {
    return formats.filter((f) => f.mimeType?.includes("audio") && f.url);
  }

  /**
   * Sort formats by quality (best first)
   */
  sortByQuality(formats: Format[]): Format[] {
    return [...formats].sort((a, b) => {
      // Video: sort by resolution then bitrate
      if (a.height && b.height) {
        if (a.height !== b.height) return b.height - a.height;
      }
      // Audio or same resolution: sort by bitrate
      return b.bitrate - a.bitrate;
    });
  }
}

File: internal/youtube/format_selector.go
package youtube

import (
	"fmt"
	"regexp"
	"strings"
)

var codecRegex = regexp.MustCompile(`codecs="?([^",]+)`)

// Video codec priority patterns (higher index = better).
var videoCodecPatterns = []struct {
	pattern *regexp.Regexp
	score   int
}{
	{regexp.MustCompile(`^vp9\.2`), 5},
	{regexp.MustCompile(`^vp9`), 4},
	{regexp.MustCompile(`^vp09`), 4},
	{regexp.MustCompile(`^av01`), 3},
	{regexp.MustCompile(`^avc1`), 2},
	{regexp.MustCompile(`^h264`), 1},
}

// Audio codec priority patterns.
var audioCodecPatterns = []struct {
	pattern *regexp.Regexp
	score   int
}{
	{regexp.MustCompile(`^opus`), 4},
	{regexp.MustCompile(`^mp4a\.40\.5`), 3},
	{regexp.MustCompile(`^mp4a\.40\.2`), 2},
	{regexp.MustCompile(`^mp4a`), 1},
}

// SelectedFormats contains the best video and audio formats.
type SelectedFormats struct {
	Video *Format
	Audio *Format
}

// SelectBestFormats selects the best video and audio formats based on configuration.
// Use SelectBestFormatsWithLogger to get debug logging of format selection.
func SelectBestFormats(formats []Format, maxResolution int, prefer60fps bool) SelectedFormats {
	return selectBestFormatsImpl(formats, maxResolution, prefer60fps, nil)
}

// SelectBestFormatsWithLogger selects formats with debug logging (matching TS formatSelector).
func SelectBestFormatsWithLogger(formats []Format, maxResolution int, prefer60fps bool, logger interface {
	Debug(msg string, args ...any)
}) SelectedFormats {
	return selectBestFormatsImpl(formats, maxResolution, prefer60fps, logger)
}

func selectBestFormatsImpl(formats []Format, maxResolution int, prefer60fps bool, logger interface {
	Debug(msg string, args ...any)
}) SelectedFormats {
	var bestVideo, bestAudio *Format
	var bestVideoCodecScore, bestAudioCodecScore int

	// Cache codec scores by mimeType to avoid redundant regex matching
	codecScoreCache := make(map[string]int)
	cachedVideoCodecScore := func(mimeType string) int {
		if score, ok := codecScoreCache[mimeType]; ok {
			return score
		}
		score := scoreVideoCodec(extractCodec(mimeType))
		codecScoreCache[mimeType] = score
		return score
	}
	cachedAudioCodecScore := func(mimeType string) int {
		key := "a:" + mimeType
		if score, ok := codecScoreCache[key]; ok {
			return score
		}
		score := scoreAudioCodec(extractCodec(mimeType))
		codecScoreCache[key] = score
		return score
	}

	// Log available video formats (matching TS FormatSelector debug output)
	if logger != nil {
		var available []string
		for i := range formats {
			f := &formats[i]
			if strings.Contains(f.MimeType, "video") && f.URL != "" {
				w, h, fps := 0, 0, 0
				if f.Width != nil {
					w = *f.Width
				}
				if f.Height != nil {
					h = *f.Height
				}
				if f.Fps != nil {
					fps = *f.Fps
				}
				available = append(available, fmt.Sprintf("itag %d (%dx%d@%dfps, %dbps)", f.Itag, w, h, fps, f.Bitrate))
			}
		}
		if len(available) > 0 {
			logger.Debug(fmt.Sprintf("[FormatSelector] Available video formats with URLs: %s", strings.Join(available, ", ")))
		}
	}

	for i := range formats {
		f := &formats[i]
		mimeType := f.MimeType

		if strings.Contains(mimeType, "video") {
			if f.URL == "" {
				continue
			}

			maxDim := f.MaxDimension()
			if maxDim > maxResolution {
				continue
			}

			if bestVideo == nil {
				bestVideo = f
				bestVideoCodecScore = cachedVideoCodecScore(f.MimeType)
				continue
			}

			bestMaxDim := bestVideo.MaxDimension()

			// Prefer higher resolution
			if maxDim > bestMaxDim {
				bestVideo = f
				bestVideoCodecScore = cachedVideoCodecScore(f.MimeType)
				continue
			}

			if maxDim == bestMaxDim {
				fFps := 30
				if f.Fps != nil {
					fFps = *f.Fps
				}
				bestFps := 30
				if bestVideo.Fps != nil {
					bestFps = *bestVideo.Fps
				}

				// FPS tiebreaker
				if fFps != bestFps {
					if prefer60fps && fFps > bestFps {
						bestVideo = f
					} else if !prefer60fps && fFps < bestFps {
						bestVideo = f
					}
					continue
				}

				// Codec score tiebreaker
				fCodec := cachedVideoCodecScore(f.MimeType)
				if fCodec != bestVideoCodecScore {
					if fCodec > bestVideoCodecScore {
						bestVideo = f
						bestVideoCodecScore = fCodec
					}
					continue
				}

				// Lower bitrate wins (better compression)
				if f.Bitrate < bestVideo.Bitrate {
					bestVideo = f
				} else if f.Bitrate == bestVideo.Bitrate {
					// Lower auth level wins
					fAuth := authLevelOf(f)
					bestAuth := authLevelOf(bestVideo)
					if fAuth < bestAuth {
						bestVideo = f
					}
				}
			}
		} else if strings.Contains(mimeType, "audio") {
			if f.URL == "" {
				continue
			}

			if bestAudio == nil {
				bestAudio = f
				bestAudioCodecScore = cachedAudioCodecScore(f.MimeType)
				continue
			}

			// Codec score
			fCodec := cachedAudioCodecScore(f.MimeType)
			if fCodec != bestAudioCodecScore {
				if fCodec > bestAudioCodecScore {
					bestAudio = f
					bestAudioCodecScore = fCodec
				}
				continue
			}

			// Higher bitrate wins
			if f.Bitrate > bestAudio.Bitrate {
				bestAudio = f
			} else if f.Bitrate == bestAudio.Bitrate {
				fAuth := authLevelOf(f)
				bestAuth := authLevelOf(bestAudio)
				if fAuth < bestAuth {
					bestAudio = f
				}
			}
		}
	}

	if logger != nil {
		if bestVideo != nil {
			w, h := 0, 0
			if bestVideo.Width != nil {
				w = *bestVideo.Width
			}
			if bestVideo.Height != nil {
				h = *bestVideo.Height
			}
			logger.Debug(fmt.Sprintf("[FormatSelector] best video: itag=%d %s %dx%d %dbps (source=%s)",
				bestVideo.Itag, extractCodec(bestVideo.MimeType),
				w, h,
				bestVideo.Bitrate, bestVideo.Source))
		} else {
			logger.Debug("[FormatSelector] no suitable video format found")
		}
		if bestAudio != nil {
			logger.Debug(fmt.Sprintf("[FormatSelector] best audio: itag=%d %s %dbps (source=%s)",
				bestAudio.Itag, extractCodec(bestAudio.MimeType),
				bestAudio.Bitrate, bestAudio.Source))
		} else {
			logger.Debug("[FormatSelector] no suitable audio format found")
		}
	}

	return SelectedFormats{Video: bestVideo, Audio: bestAudio}
}

// SelectWithOptions selects formats with optional manual itag override.
func SelectWithOptions(formats []Format, maxResolution int, prefer60fps bool, videoItag, audioItag *int) SelectedFormats {
	var video, audio *Format

	// Manual video selection
	if videoItag != nil && *videoItag != -1 {
		for i := range formats {
			if formats[i].Itag == *videoItag && strings.Contains(formats[i].MimeType, "video") && formats[i].URL != "" {
				video = &formats[i]
				break
			}
		}
	}

	// Manual audio selection
	if audioItag != nil && *audioItag != -1 {
		for i := range formats {
			if formats[i].Itag == *audioItag && strings.Contains(formats[i].MimeType, "audio") && formats[i].URL != "" {
				audio = &formats[i]
				break
			}
		}
	}

	// Explicit skip
	skipVideo := videoItag != nil && *videoItag == -1
	skipAudio := audioItag != nil && *audioItag == -1

	// Auto-select for unresolved
	auto := SelectBestFormats(formats, maxResolution, prefer60fps)
	if video == nil && !skipVideo {
		video = auto.Video
	}
	if audio == nil && !skipAudio {
		audio = auto.Audio
	}

	return SelectedFormats{Video: video, Audio: audio}
}

func extractCodec(mimeType string) string {
	m := codecRegex.FindStringSubmatch(mimeType)
	if m == nil {
		return ""
	}
	return m[1]
}

func scoreVideoCodec(codec string) int {
	for _, p := range videoCodecPatterns {
		if p.pattern.MatchString(codec) {
			return p.score
		}
	}
	return 0
}

func scoreAudioCodec(codec string) int {
	for _, p := range audioCodecPatterns {
		if p.pattern.MatchString(codec) {
			return p.score
		}
	}
	return 0
}

func authLevelOf(f *Format) int {
	if f.AuthLevel != nil {
		return *f.AuthLevel
	}
	return 999
}
